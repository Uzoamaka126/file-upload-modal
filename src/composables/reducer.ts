import { ref, reactive } from 'vue';
import { ReducerParam, InitialState } from './types';

export function useReducer(reducer: any, initialState: InitialState): ReducerParam {
    const state = reactive(initialState);
    const stateRef = ref(state);

    const dispatch = (action: any) => {
        stateRef.value = reducer(stateRef.value, action);
    };

    return [stateRef, dispatch];
};