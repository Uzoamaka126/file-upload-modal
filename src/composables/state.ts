import { readonly, ref } from 'vue';

export function useState(initialState: any) {
  const state = ref(initialState);

  const setState = (newState: any) => {
    state.value = newState;
  };

  const readonlyState = readonly(state);
  
  return [readonlyState, setState];
}