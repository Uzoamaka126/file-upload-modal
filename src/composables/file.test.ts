import { expect, it, describe } from 'vitest';
import { shallowMount } from '@vue/test-utils'
import UploadModalVue from '../components/UploadModal.vue';
import { useReducer } from './reducer';
import { useState } from './state';
import { ref } from 'vue';

describe('test composables', () => {
    it('should mount component', () => {
        const wrapper = shallowMount(UploadModalVue, {
          global: {
            stubs: ['CloudIcon'],
          },
        });
        expect(wrapper.html()).toMatchSnapshot()
    });

    it('return a new state and dispatch with useReducer call', async () => {
        const initialState = ref(0)

        function sampleReducer(state: any, action: any) {
            switch(action.type) {
                case 'INCREMENT':
                    return state + 1;
                case 'DECREMENT':
                    return state - 1;
                default:
                    return state;
            }
        };

        const [state, dispatch] = useReducer(sampleReducer, initialState.value);
        
        expect(state.value).toBe(0);

        await dispatch!({ type: 'INCREMENT' });
      
        expect(state.value).toBe(1)
    })

    it('return a readonly state and setter function with useState', async () => {
        const [state, setState] = useState(0);
        
        expect(state.value).toBe(0);

        await setState(1)
      
        expect(state.value).toBe(1);
    })
});