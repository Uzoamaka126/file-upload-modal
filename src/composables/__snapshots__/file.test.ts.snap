// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mount component > should mount component 1`] = `
"<transition-stub data-v-d3ad3f7a="" appear="false" persisted="false" css="true">
  <!--v-if-->
</transition-stub>"
`;

exports[`mount component 1`] = `
"<transition-stub data-v-d3ad3f7a="" appear="false" persisted="false" css="true">
  <!--v-if-->
</transition-stub>"
`;

exports[`test composables > should mount component 1`] = `
"<transition-stub data-v-d3ad3f7a="" appear="false" persisted="false" css="true">
  <!--v-if-->
</transition-stub>"
`;
